### PR TITLE
WOC: Implement utm* params of a login UserAction stored in the DB feature #2212 #2359

### PR DESCRIPTION
**Pull request by unclebean**
##### Description
The 3rd times create same pull request...

This PR is for saving utm* params to DB for data analysis, relevant story is [2212](https://github.com/gitcoinco/web/issues/2212). 
Whenever user access gitcoin with "utm_medum, utm_source, utm_campaign" will store them in cookie, after user login or user logout will store to dashboard_useraction table.

##### Checklist

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)

UserAction data model.

##### Testing

Test with below 3 scenarios.

**Scenario:** Record utm* in Cookie
**Give:** utm_medum, utm_source, utm_campaign in URL
**When:** User open gitcoin
**Then:** Should record utm* in Cookie

**Scenario:** Record utm* in DB
**Give:** utm* in Cookie
**When:** User click "LOG IN" button
**Then:** Should record utm* in DB

**Scenario:** Record utm* in DB
**Give:** utm* in Cookie
**When:** User click "Log Out" button
**Then:** Should record utm* in DB

### CODE REVIEW

**Comment by frankchen07** [comment]
@owocki I believe this is ready for a reviewer.

**Comment by unclebean** [comment]
hi @mbeacom 
May I know the status of merging this PR? I want to pick another task once this one is done.

File: app/assets/v2/js/base.js [hunk 1]
```diff
@@ -16,6 +16,23 @@ $(document).ready(function() {
...
+  var record_campaign_to_cookie = function() {
+    var paramsStr = decodeURIComponent(window.location.search.substring(1));
+
+    paramsStr.split('&')
+      .map((paramStr) => {
+        var [ key, value ] = paramStr.split('=');
+
+        return {key, value};
+      })
+      .filter((param) => [ 'utm_medium', 'utm_source', 'utm_campaign' ].indexOf(param.key) !== -1)
```
Comment (timborden) [inline app/assets/v2/js/base.js]: Nit - can we use [`includes`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) here?

```suggestion
      .filter((param) => [ 'utm_medium', 'utm_source', 'utm_campaign' ].includes(param.key))
```

File: app/dashboard/tests/test_dashboard_utils.py [hunk 2]
```diff
@@ -73,3 +78,35 @@ def test_humanize_event_name():
...
+        create_user_action(None, 'Login', request)
+        mockUserAction.create.assert_called_once_with(action='Login', metadata={}, user=None,
+                                                      utm={'utm_source': 'test source',
+                                                           'utm_medium': 'test medium',
+                                                           'utm_campaign': 'test campaign'})
+
+    @staticmethod
+    @patch('dashboard.utils.UserAction.objects')
+    def test_create_user_action_without_cookie(mockUserAction):
+        """Test the giving utm* in cookie should store in DB."""
```
Comment (timborden) [inline app/dashboard/tests/test_dashboard_utils.py]: Nit - comment needs updating

File: app/dashboard/tests/test_dashboard_utils.py [hunk 3]
```diff
@@ -73,3 +78,35 @@ def test_humanize_event_name():
...
+        request = RequestFactory().get('/login')
+        create_user_action(None, 'Login', request)
+        mockUserAction.create.assert_called_once_with(action='Login', metadata={}, user=None)
+
+    @staticmethod
+    @patch('dashboard.utils.UserAction.objects')
+    def test_create_user_action_campaign_not_json(mockUserAction):
+        """Test the giving utm* in cookie should store in DB."""
```
Comment (timborden) [inline app/dashboard/tests/test_dashboard_utils.py]: Nit - comment needs updating

File: app/dashboard/tests/test_dashboard_utils.py [hunk 4]
```diff
@@ -73,3 +78,35 @@ def test_humanize_event_name():
...
+    @patch('dashboard.utils.UserAction.objects')
+    def test_create_user_action_without_cookie(mockUserAction):
+        """Test the giving utm* in cookie should store in DB."""
+        request = RequestFactory().get('/login')
+        create_user_action(None, 'Login', request)
+        mockUserAction.create.assert_called_once_with(action='Login', metadata={}, user=None)
+
+    @staticmethod
+    @patch('dashboard.utils.UserAction.objects')
+    def test_create_user_action_campaign_not_json(mockUserAction):
```
Comment (timborden) [inline app/dashboard/tests/test_dashboard_utils.py]: How is this test different then `test_create_user_action_with_cookie`?

File: app/assets/v2/js/base.js (hunk 1 above)
Comment (unclebean) [reply app/assets/v2/js/base.js]: you no need to support IE?

File: app/dashboard/tests/test_dashboard_utils.py (hunk 4 above)
Comment (unclebean) [reply app/dashboard/tests/test_dashboard_utils.py]: I want to test partial value in cookie scenario...

File: app/assets/v2/js/base.js (hunk 1 above)
Comment (timborden) [reply app/assets/v2/js/base.js]: Hmm....I can see [babel in the devDependencies](https://github.com/gitcoinco/web/blob/master/package.json#L24)...and it looks like it's used in the [Webpack config](https://github.com/gitcoinco/web/blob/master/webpack.config.js#L35) (but I'm not a Webpack expert, TBH) so I think IE should be supported.

File: app/dashboard/tests/test_dashboard_utils.py (hunk 4 above)
Comment (timborden) [reply app/dashboard/tests/test_dashboard_utils.py]: Gotcha.

Does it make sense to rename the test method to something like `test_create_user_action_with_partial_cookie`?

File: app/assets/v2/js/base.js (hunk 1 above)
Comment (timborden) [reply app/assets/v2/js/base.js]: Ahhh....fair enough. Suppose babel config could be updated...but likely out of scope for this PR.

**Pull request merged**